### ISSUE TITLE
`adaptive_max_pool3d_backward` and `fractional_max_pool{2,3}d_backward` read a non-contiguous `indices` tensor as if it were contiguous: silently wrong gradients, reached from `torch.compile` with `channels_last_3d` inputs

### ISSUE THREAD

**Issue by cuiliaomei-beep** [pytorch/pytorch#197888]
### 🐛 Describe the bug

**Summary.** For `F.adaptive_max_pool3d` on a `channels_last_3d` input, eager and `aot_eager` compute the same input gradient, and Inductor computes a different one: the forward values and indices are identical, but `backward()` scatters the gradient to the wrong positions (max difference 282 in the example below; torch 2.14.0 and nightly). A 3-D CNN that follows the memory-format recommendation and is then compiled trains with silently wrong gradients through every adaptive max pool. Because the CPU kernels of `adaptive_max_pool3d_backward`, `fractional_max_pool2d_backward`, and `fractional_max_pool3d_backward` make `grad_output` contiguous but read `indices` through its raw data pointer, any caller that passes non-contiguous indices gets a wrong result; eager autograd never does, Inductor does as soon as it lays the saved indices out in the input's memory format.

User-level reproduction (needs Inductor):

```python
import torch, torch.nn.functional as F

torch.manual_seed(0)
f = lambda t: F.adaptive_max_pool3d(t, (2, 3, 5))
x = torch.randperm(1 * 4 * 4 * 6 * 5).float().reshape(1, 4, 4, 6, 5)
for name, xin in (("contiguous", x), ("channels_last_3d", x.contiguous(memory_format=torch.channels_last_3d))):
    grads = []
    for backend in (None, "aot_eager", "inductor"):
        torch._dynamo.reset()
        t = xin.detach().clone(memory_format=torch.preserve_format).requires_grad_()
        out = (f if backend is None else torch.compile(f, backend=backend))(t)
        (out * torch.arange(1.0, out.numel() + 1).reshape(out.shape)).sum().backward()
        grads.append(t.grad.clone())
    print(f"{name:17s} aot_eager == eager: {torch.equal(grads[0], grads[1])}   inductor == eager: {torch.equal(grads[0], grads[2])}"
          f"   max diff {(grads[0] - grads[2]).abs().max().item():.0f}")
```

```
contiguous        aot_eager == eager: True   inductor == eager: True   max diff 0
channels_last_3d  aot_eager == eager: True   inductor == eager: False   max diff 282
```

Kernel-level reproduction (eager only, no compiler involved): the same values in `indices`, once contiguous and once with `channels_last` strides.

```python
import torch, torch.nn.functional as F

torch.manual_seed(0)
A = torch.ops.aten
x3 = torch.randperm(240).float().reshape(1, 4, 4, 3, 5)
x3w = torch.cat([x3, x3], 3)
x2 = torch.randperm(120).float().reshape(1, 4, 6, 5)
rs2, rs3 = torch.full((1, 4, 2), 0.5), torch.full((1, 4, 3), 0.5)
cases = []
v, i = F.adaptive_max_pool3d(x3, (2, 3, 5), return_indices=True)
cases.append(("adaptive_max_pool3d_backward", lambda go, idx: A.adaptive_max_pool3d_backward(go, x3, idx), v, i, torch.channels_last_3d))
v, i = F.adaptive_max_pool2d(x2, (3, 5), return_indices=True)
cases.append(("adaptive_max_pool2d_backward", lambda go, idx: A.adaptive_max_pool2d_backward(go, x2, idx), v, i, torch.channels_last))
v, i = F.max_pool3d(x3, 2, 1, return_indices=True)
cases.append(("max_pool3d_with_indices_backward", lambda go, idx: A.max_pool3d_with_indices_backward(go, x3, [2, 2, 2], [1, 1, 1], [0, 0, 0], [1, 1, 1], False, idx), v, i, torch.channels_last_3d))
v, i = F.fractional_max_pool2d(x2, 2, output_size=(3, 3), return_indices=True, _random_samples=rs2)
cases.append(("fractional_max_pool2d_backward", lambda go, idx: A.fractional_max_pool2d_backward(go, x2, [2, 2], [3, 3], idx), v, i, torch.channels_last))
v, i = F.fractional_max_pool3d(x3w, 2, output_size=(2, 2, 2), return_indices=True, _random_samples=rs3)
cases.append(("fractional_max_pool3d_backward", lambda go, idx: A.fractional_max_pool3d_backward(go, x3w, [2, 2, 2], [2, 2, 2], idx), v, i, torch.channels_last_3d))
for name, fn, v, idx, fmt in cases:
    go = torch.arange(1.0, v.numel() + 1).reshape(v.shape)
    idx_nc = idx.contiguous(memory_format=fmt)                      # equal values, channels_last strides
    same = torch.equal(fn(go, idx), fn(go, idx_nc))
    print(f"{name:34s} indices strides {str(tuple(idx_nc.stride())):22s} same gradient as with contiguous indices: {same}")
```

```
adaptive_max_pool3d_backward       indices strides (120, 1, 60, 20, 4)    same gradient as with contiguous indices: False
adaptive_max_pool2d_backward       indices strides (60, 1, 20, 4)         same gradient as with contiguous indices: True
max_pool3d_with_indices_backward   indices strides (96, 1, 32, 16, 4)     same gradient as with contiguous indices: True
fractional_max_pool2d_backward     indices strides (36, 1, 12, 4)         same gradient as with contiguous indices: False
fractional_max_pool3d_backward     indices strides (32, 1, 16, 8, 4)      same gradient as with contiguous indices: False
```

`adaptive_max_pool2d_backward` and `max_pool{2,3}d_with_indices_backward` normalize `indices` and are correct; the other three are not. Passing a channels_last `grad_output` or `self` to the affected kernels is handled correctly, only `indices` is read without `.contiguous()`.

**Possible fix.** In `aten/src/ATen/native/AdaptiveMaxPooling3d.cpp`, `FractionalMaxPool2d.cpp`, and `FractionalMaxPool3d.cpp` (backward, CPU; the CUDA counterparts are worth the same look), take `indices.contiguous()` next to the existing `gradOutput.contiguous()`; the 2-D adaptive kernel and the `max_pool{2,3}d_with_indices_backward` kernels are correct in the table above, so they evidently normalize the argument already. Independently, Inductor's fallback for these ops could `require_contiguous` the `indices` argument. This belongs to the class tracked in #195547 (a precondition the kernel relies on but never establishes).

### Error logs

No error is raised; the gradients differ as shown above.

### Versions

```
PyTorch version: 2.14.0+cpu
OS: Microsoft Windows 11 Pro for Workstations (10.0.26200 64-bit)
MSVC: Microsoft (R) C/C++ Optimizing Compiler Version 19.44.35228 for x64
Python version: 3.14.7 (64-bit runtime)
Is CUDA available: False
CPU: 11th Gen Intel(R) Core(TM) i9-11900T @ 1.50GHz
[pip3] numpy==2.5.2
[pip3] torch==2.14.0+cpu
```

Same output on nightly (2.15.0.dev20260911 on Windows, 2.15.0.dev20260920 on Linux) and on Linux with 2.14.0. On CUDA (Tesla T4, 2.14.0+cu130) the compiled gradient matches eager, so only the CPU kernels are affected.

cc @ezyang @gchanan @kadeng @msaroufim @jamesr66a @mikaylagawarecki @chauhang @penguinwu

**Comment by chinmayirm** [comment]
I’d like to work on this.

These backward kernels scatter `grad_output` to the argmax locations stored in `indices`. They already do `gradOutput.contiguous()`, then take `indices.const_data_ptr<int64_t>()` and walk it as packed NCHW (`ind_p_d[ot * osizeH * osizeW + oh * osizeW + ow]`). That arithmetic is only valid if `indices` has unit inner stride. If it is `channels_last` / `channels_last_3d`, the same values sit at different offsets, `maxp` is the wrong voxel, and the gradient is silently wrong. Forward still matches because the forward kernel wrote the indices; only backward misreads them.

Eager autograd always saves contiguous indices, so this never shows up there. Inductor can keep the input memory format on the saved indices, which is how `torch.compile` reaches it.

`adaptive_max_pool2d_backward` is the existence proof that this is just a missing `contiguous()`: `AdaptiveMaxPoolKernel.cpp` already does `indices.contiguous()`. A 3d kernel in that file does too, but `TORCH_IMPL_FUNC` in `AdaptiveMaxPooling3d.cpp` never calls it — it still uses the old templates. CUDA `adaptive_max_pool3d_backward` in `AdaptiveMaxPooling3d.cu` has the same `data_ptr` walk. CUDA fractional uses `packed_accessor`, which respects strides, so those should be left alone.

Proposed approach
`auto indices_ = indices.contiguous();` next to `gradOutput_` in `AdaptiveMaxPooling3d.cpp`, `FractionalMaxPool2d.cpp`, `FractionalMaxPool3d.cpp`, and `AdaptiveMaxPooling3d.cu`. Use `indices_` for `data_ptr`. I would not rewire 3d onto the unused dispatched kernel here; `contiguous()` is the 2d fix, and it is a no-op when indices are already contiguous.

Tests: same index values with `channels_last` / `channels_last_3d` strides vs contiguous — gradients must match for the three CPU ops and CUDA `adaptive_max_pool3d`. Also `torch.compile` + `adaptive_max_pool3d` on a `channels_last_3d` input: backward matches eager.

I would be happy to raise a PR if assigned. Thanks!
